Drag snapshot paints unrelated content

nullhook opened the pull request:
#### a3e5281daec2d26daf7fdf385a0fd213437e814c
<pre>
Drag snapshot paints unrelated content
<a href="https://bugs.webkit.org/show_bug.cgi?id=237446">https://bugs.webkit.org/show_bug.cgi?id=237446</a>
<a href="https://rdar.apple.com/90120656">rdar://90120656</a>

Reviewed by Simon Fraser.

When using setDragImage with a fixed-position element,
the drag preview bitmap incorrectly included content
from other stacking contexts.

During painting, we did not properly filter
layers based on subtreePaintRoot, causing unrelated
stacking context layers to be composited
into the dragged element&apos;s snapshot.

This fix ensures that only content layers within
the same stacking context are painted.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintLayer):
(WebCore::RenderLayer::paintLayerContents):

Canonical link: <a href="https://commits.webkit.org/296117@main">https://commits.webkit.org/296117@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9afa2b9e444fee6d63b026d5c0178a47fdf106b1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107439 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27121 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17528 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112653 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/57975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109403 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27804 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35621 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81559 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/57975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110368 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22021 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96829 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61935 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21461 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14960 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57419 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91391 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14995 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115754 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34506 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25430 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90597 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34881 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93079 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90336 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23032 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35249 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13030 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/30249 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34427 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39968 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34173 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37528 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35834 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3743,6 +3743,17 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
+    if (shouldPaintContent && paintingInfo.subtreePaintRoot) {
+        RenderLayer* subtreeRootLayer = paintingInfo.subtreePaintRoot->enclosingLayer();
+
+        if (subtreeRootLayer && subtreeRootLayer->isStackingContext()) {
```
This has a number of problems. First, it needs to work for nested stacking contexts; you need to paint descendants in cases like this:

```
<div style="opacity: 0.9"> <!-- snapshot node -->
  <div>
    <div style="opacity: 0.9">
       Is this included?
   </div>
  </div>
</div>

```

Second, I think it *also* needs to ask question about containing block. For example, you probably don't want to paint the abspos here:

```
<div style="opacity: 0.9"> <!-- snapshot node -->
  <div>
    <div style="position: absolute">
       Is this excluded?
   </div>
  </div>
</div>

```

We need to find a way to make tests for this. Can we get snapshots in JS and paint to canvas, for testing?

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3743,6 +3743,34 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
+        if (subtreeRootLayer) {
+            bool isLayerInSubtree = false;
+
+            for (auto* layer = this; layer; layer = layer->parent()) {
```
Use `RenderLayer::isDescendantOf()`

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3743,6 +3743,34 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
+            }
+
+            if (isLayerInSubtree) {
+                RenderBlock* containingBlock = renderer().containingBlock();
```
I think the right question here is "Is paintingInfo.subtreePaintRoot a containing block ancestor", which can be answered by `RenderBlock::isContainingBlockAncestorFor()`.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3743,6 +3743,24 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
+
+            if (isLayerInSubtree) {
+                if (paintingInfo.subtreePaintRoot != &renderer()) {
+                    if (auto* rootAsBlock = dynamicDowncast<RenderBlock>(paintingInfo.subtreePaintRoot)) {
```
CheckedPtr

nullhook committed:
Drag snapshot paints unrelated content
https://bugs.webkit.org/show_bug.cgi?id=237446
rdar://90120656

Reviewed by Simon Fraser.

When using setDragImage with a fixed-position element,
the drag preview bitmap incorrectly included content
from other stacking contexts.

During painting, we did not properly filter
layers based on subtreePaintRoot, causing unrelated
stacking context layers to be composited
into the dragged element's snapshot.

This fix ensures that only content layers within
the same stacking context are painted.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintLayer):
(WebCore::RenderLayer::paintLayerContents):

Canonical link: https://commits.webkit.org/296117@main
